proparse/src/main/java/org/prorefactor/proparse/antlr4/TokenList.java

```diff
@@ -0,0 +1,191 @@
+import org.antlr.v4.runtime.Token;
+import org.antlr.v4.runtime.TokenFactory;
+import org.antlr.v4.runtime.TokenSource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * List of tokens for input to the parser. This class is responsible for gathering a list of tokens from the PostLexer,
+ * and examining that list for fine-tuning before sending it on to the parser.
+ * 
+ * The complete tool chain is:<ol>
+ * <li>Lexer 
+ * <li>PostLexer
+ * <li>TokenList
+ * <li>parser
+ * </ol>
+ * 
+ * TODO The temporary list should be avoided
+ */
+public class TokenList implements TokenSource {
+  private static final Logger LOGGER = LoggerFactory.getLogger(TokenList.class);
+
+  private final TokenSource tokenStream;
+  private final List<ProToken> list = new ArrayList<>();
+
+  private int currentPosition = 0;
+  private ProToken currentToken;
+  private boolean initialized = false;
+
+  TokenList(TokenSource input) {
+    this.tokenStream = input;
+  }
+
+  private void build()  {
+    LOGGER.trace("Entering TokenList#build()");
+    int zz = 0;
+    for (;;) {
+      ProToken nextToken = (ProToken) tokenStream.nextToken();
+      nextToken.setTokenIndex(zz++);
+      list.add(nextToken);
+      if (nextToken.getType() == PreprocessorParser.OBJCOLON)
+        reviewObjcolon();
+      if (nextToken.getType() == PreprocessorParser.EOF)
+        break;
+    }
+    LOGGER.trace("Exiting TokenList#build() - {} tokens", list.size());
+  }
+
+  /**
+   * Review the token list at an OBJCOLON token.
+   * 
+   * This is the reason this class was created in the first place. If we have an OBJCOLON, what comes before it has to
+   * be one of a few things:
+   * <ul>
+   * <li>a system handle,
+   * <li>a handle expression,
+   * <li>an Object reference expression, or
+   * <li>a type name (class name) for a static member reference
+   * </ul>
+   * <p>
+   * A type name can be pretty much anything, even a reserved keyword. It can also be a qualified class name, such as
+   * com.joanju.Foo.
+   * <p>
+   * This method attempts to resolve the following problem: Because of static class member references, a class name can
+   * be the first token in an expression. Class names can be composed of reserved keywords. This means that a reserved
+   * keyword could be the first piece of an expression, and this completely breaks the lookahead in the ANTLR generated
+   * parser. So, here, we look for OBJCOLON, and make sure that what comes before it is a system handle, an ID, or a
+   * non-reserved keyword.
+   * <p>
+   * A NAMEDOT token is a '.' followed by anything other than whitespace. If the OBJCOLON is proceeded by a NAMEDOT pair
+   * (NAMEDOT followed by anything), then we convert all of the NAMEDOT pairs to NAMEDOT-ID pairs. Otherwise, if the
+   * OBJCOLON is proceeded by any reserved keyword other than a systemhandlename, then we change that token's type to
+   * ID.
+   */
+  private void reviewObjcolon() {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) The Cyclomatic Complexity of this method "reviewObjcolon" is 13 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://sonar.riverside-software.fr/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)